lib/charms/layer/apache_zeppelin.py

```diff
@@ -144,18 +204,50 @@ def start(self):
             zeppelin_home = self.dist_config.path('zeppelin')
             # chdir here because things like zepp tutorial think ZEPPELIN_HOME
             # is wherever the daemon was started from.
-            os.chdir(zeppelin_home)
-            utils.run_as('ubuntu',
-                         '{}/bin/zeppelin-daemon.sh'.format(zeppelin_home),
-                         '--config', zeppelin_conf,
-                         'start')
+            with host.chdir(zeppelin_home):
+                utils.run_as('ubuntu',
+                             '{}/bin/zeppelin-daemon.sh'.format(zeppelin_home),
+                             '--config', zeppelin_conf,
+                             'start')
+            # wait up to 30s for server to start responding, lest API requests fail
+            self.wait_for_api(30)
+
+    def check_connect(self, addr, port):
+        try:
+            with socket.create_connection((addr, port), timeout=10):
+                return True
+        except OSError:
+            return False
+
+    def wait_for_api(self, timeout):
+        start = time.time()
+        while time.time() - start < timeout:
+            if self.check_connect('localhost', self.dist_config.port('zeppelin')):
+                return True
+            time.sleep(2)
+        raise utils.TimeoutError('Timed-out waiting for connection to Zeppelin')
+
+    def wait_for_stop(self, timeout):
+        start = datetime.now()
+        while utils.jps("zeppelin"):
+            time.sleep(1)
+            if datetime.now() - start > timeout:
```

Comment:
We get this here. We probably miss a .seconds like (datetime.now() - start).seconds

```
File "/var/lib/juju/agents/unit-apache-zeppelin-0/charm/hooks/spark-relation-departed", line 19, in <module>
2016-10-03 18:30:04 INFO spark-relation-departed     main()
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/__init__.py", line 78, in main
2016-10-03 18:30:04 INFO spark-relation-departed     bus.dispatch()
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 434, in dispatch
2016-10-03 18:30:04 INFO spark-relation-departed     _invoke(other_handlers)
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 417, in _invoke
2016-10-03 18:30:04 INFO spark-relation-departed     handler.invoke()
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 291, in invoke
2016-10-03 18:30:04 INFO spark-relation-departed     self._action(*args)
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/charm/reactive/zeppelin.py", line 51, in stop_zeppelin
2016-10-03 18:30:04 INFO spark-relation-departed     zepp.stop()
2016-10-03 18:30:04 INFO spark-relation-departed   File "lib/charms/layer/apache_zeppelin.py", line 255, in stop
2016-10-03 18:30:04 INFO spark-relation-departed     self.wait_for_stop(30)
2016-10-03 18:30:04 INFO spark-relation-departed   File "lib/charms/layer/apache_zeppelin.py", line 246, in wait_for_stop
2016-10-03 18:30:04 INFO spark-relation-departed     if datetime.now() - start > timeout:
2016-10-03 18:30:04 INFO spark-relation-departed TypeError: unorderable types: datetime.timedelta() > int()

```